core/chains/evm/txm/storage/inmemory_store.go

```diff
@@ -0,0 +1,333 @@
+	pruneSubset           = 3
+)
+
+type InMemoryStore struct {
+	sync.RWMutex
+	lggr      logger.Logger
+	txIDCount uint64
+	address   common.Address
+	chainID   *big.Int
+
+	UnstartedTransactions   []*types.Transaction
+	UnconfirmedTransactions map[uint64]*types.Transaction
+	ConfirmedTransactions   map[uint64]*types.Transaction
+	FatalTransactions       []*types.Transaction
+
+	Transactions map[uint64]*types.Transaction
+}
+
+func NewInMemoryStore(lggr logger.Logger, address common.Address, chainID *big.Int) *InMemoryStore {
+	return &InMemoryStore{
+		lggr:                    logger.Named(lggr, "InMemoryStore"),
+		address:                 address,
+		chainID:                 chainID,
+		UnconfirmedTransactions: make(map[uint64]*types.Transaction),
+		ConfirmedTransactions:   make(map[uint64]*types.Transaction),
+		Transactions:            make(map[uint64]*types.Transaction),
+	}
+}
+
+func (m *InMemoryStore) AbandonPendingTransactions() {
+	m.Lock()
+	defer m.Unlock()
+
+	for _, tx := range m.UnstartedTransactions {
+		tx.State = types.TxFatalError
+	}
+	m.FatalTransactions = m.UnstartedTransactions
+	m.UnstartedTransactions = []*types.Transaction{}
+
+	for _, tx := range m.UnconfirmedTransactions {
+		tx.State = types.TxFatalError
+		m.FatalTransactions = append(m.FatalTransactions, tx)
+	}
+	m.UnconfirmedTransactions = make(map[uint64]*types.Transaction)
+}
+
+func (m *InMemoryStore) AppendAttemptToTransaction(txNonce uint64, attempt *types.Attempt) error {
+	m.Lock()
+	defer m.Unlock()
+
+	tx, exists := m.UnconfirmedTransactions[txNonce]
+	if !exists {
+		return fmt.Errorf("unconfirmed tx was not found for nonce: %d - txID: %v", txNonce, attempt.TxID)
+	}
+
+	if tx.ID != attempt.TxID {
+		return fmt.Errorf("unconfirmed tx with nonce exists but attempt points to a different txID. Found Tx: %v - txID: %v", m.UnconfirmedTransactions[txNonce], attempt.TxID)
+	}
+
+	attempt.CreatedAt = time.Now()
+	attempt.ID = uint64(len(tx.Attempts)) // Attempts are not collectively tracked by the in-memory store so attemptIDs are not unique between transactions and can be reused.
+	tx.AttemptCount++
+	m.UnconfirmedTransactions[txNonce].Attempts = append(m.UnconfirmedTransactions[txNonce].Attempts, attempt.DeepCopy())
+
+	return nil
+}
+
+func (m *InMemoryStore) CountUnstartedTransactions() int {
+	m.RLock()
+	defer m.RUnlock()
+
+	return len(m.UnstartedTransactions)
+}
+
+func (m *InMemoryStore) CreateEmptyUnconfirmedTransaction(nonce uint64, gasLimit uint64) (*types.Transaction, error) {
+	m.Lock()
+	defer m.Unlock()
+
+	m.txIDCount++
+	emptyTx := &types.Transaction{
+		ID:                m.txIDCount,
+		ChainID:           m.chainID,
+		Nonce:             &nonce,
+		FromAddress:       m.address,
+		ToAddress:         common.Address{},
+		Value:             big.NewInt(0),
+		SpecifiedGasLimit: gasLimit,
+		CreatedAt:         time.Now(),
+		State:             types.TxUnconfirmed,
+	}
+
+	if _, exists := m.UnconfirmedTransactions[nonce]; exists {
```

Comment:
Same here if we need to check if the nonce exists in the confirmed map too for safety